### PR TITLE
Allow configuring cssnano

### LINKED ISSUE #1721: Parcel not respecting cssnano config
# 🐛 bug report
I have my Sass .scss files which are compiled and then I run those through autoprefixer and cssnano. The problem is that when I do this, cssnano transforms expressions like `scale(1, 0)` to `scaleY(0)` which causes issues in Safari. So I decided I'll just turn off the [`reduceTransforms`](https://cssnano.co/optimisations/reducetransforms/) option in cssnano. However even though I have done this, after running Parcel I still get the same CSS with the transformation of `scale(1, 0)` to `scaleY(0)` being done.

## 🎛 Configuration (.babelrc, package.json, cli command)

My `postcss.config.js`:

```js
module.exports = {
  plugins: [
    require('autoprefixer'),
    require('cssnano')({
      preset: ['default', {
        reduceTransforms: false
      }]
    })
  ]
};
```

And to run parcel I am doing this CLI command:

```
npx parcel build scss/main.scss -d static/css -o main.css
```

## 🤔 Expected Behavior
As above, I am expecting `static/css/main.css` to not have this transformation being applied because in my config I have specified `reduceTransforms: false`

## 😯 Current Behavior
Instead, this transformation is being applied.

## 💁 Possible Solution
I'm not sure why this is happening. Intentionally putting invalid syntax in the postcss.config.js throws a syntax error when running parcel so it is definitely reading the postcss.config.js. Also it does look like autoprefixer is being run correctly so not sure why cssnano wouldn't.

This also doesn't happen when I pass the `--no-minify` flag but I do want minified output

## 🔦 Context
I am statically generating CSS files from SASS which are placed in the server's static files directory so I can serve them. I want to minify the code but this cssnano optimization results in an animation of the `transform:` property not working in Safari.

## 💻 Code Sample

Source code is this line:

```css
.popover {
    transform: scale(1, 0) rotateX(-90deg);
}
```

and out I get:

```css
transform: scaleY(0) rotateX(-90deg);
```

## 🌍 Your Environment

| Software         | Version(s) |
| ---------------- | ---------- |
| Parcel           | 1.8.1 (running slightly old version because otherwise babel 7 results in error)
| Node             | v8.9.3
| npm/Yarn         | 6.0.1
| Operating System | macOS 10.13.4

other versions:

 - cssnano: 4.0.1

### LINKED ISSUE #6102: Upgrade to cssnano 5 in optimizer
# 🙋 feature request

Upgrade https://github.com/parcel-bundler/parcel/blob/v2/packages/optimizers/cssnano/src/CSSNanoOptimizer.js to use https://github.com/cssnano/cssnano/releases/tag/cssnano%405.0.0

### PR DESCRIPTION
**Pull request by aminya**
# ↪️ Pull Request

- Allow configuring cssnano
- Update postcss: https://github.com/postcss/postcss/blob/main/CHANGELOG.md
- Update cssnano

Fixes #1721

Already fixed. Closes #6102 

## 🚨 Test instructions
Install cssnano-preset-advanced, then make the following file at the project root.

cssnano.config.js
```js
module.exports = {
  preset: [require("cssnano-preset-advanced")]
}
```

Build and pack this cssnano package, then install it in your project. Using pnpm you can simply do this in your package.json
```json
  "pnpm": {
    "overrides": {
      "@parcel/optimizer-cssnano": "link:../parcel/packages/optimizers/cssnano"
    }
  },
```

## ✔️ PR Todo

- [ ] Added/updated unit tests for this change
- [x] Filled out test instructions (In case there aren't any unit tests)
- [x] Included links to related issues/PRs

### CODE REVIEW

**Comment by devongovett** [comment]
Thanks. FWIW, after #6379 lands, this will need to be refactored to move the config loading into the plugin's `loadConfig` method, which will handle properly invalidating the cache when the config changes.

**Comment by aminya** [comment]
> Thanks. FWIW, after #6379 lands, this will need to be refactored to move the config loading into the plugin's `loadConfig` method, which will handle properly invalidating the cache when the config changes.

Do you mean this?

```diff
diff --git a/packages/optimizers/cssnano/src/CSSNanoOptimizer.js b/packages/optimizers/cssnano/src/CSSNanoOptimizer.js
index bb96edebb..1ff1a250f 100644
--- a/packages/optimizers/cssnano/src/CSSNanoOptimizer.js
+++ b/packages/optimizers/cssnano/src/CSSNanoOptimizer.js
@@ -9,11 +9,21 @@ import {loadConfig} from '@parcel/utils';
 import path from 'path';

 export default (new Optimizer({
+  async loadConfig({ options }) {
+    return ((await loadConfig(
+      options.inputFS,
+      path.join(options.entryRoot, 'index.css'),
+      ['.cssnanorc ', 'cssnano.config.json', 'cssnano.config.js'],
+      options.projectRoot,
+    )) ?? {}: CSSNanoOptions);
+  }
+
   async optimize({
     bundle,
     contents: prevContents,
     getSourceMapReference,
     map: prevMap,
+    config,
     options,
   }) {
     if (!bundle.env.shouldOptimize) {
@@ -26,14 +36,7 @@ export default (new Optimizer({
       );
     }

-    const userConfig = ((await loadConfig(
-      options.inputFS,
-      path.join(options.entryRoot, 'index.css'),
-      ['.cssnanorc ', 'cssnano.config.json', 'cssnano.config.js'],
-      options.projectRoot,
-    )) ?? {}: CSSNanoOptions);
-
+    const result = await postcss([cssnano(config)]).process(prevContents, {
       // Suppress postcss's warning about a missing `from` property. In this
       // case, the input map contains all of the sources.
       from: undefined,
```

**Comment by mischnic** [comment]
Almost, use `config.getConfigFrom` instead of `utils.loadConfig()`: https://github.com/parcel-bundler/parcel/commit/1ba0bd1833d785c1acf6af36e0a1d429f1ce1fb6#diff-eb084320ed4553866490d232b885f8a522e1f08874f30a75ffb2ca927bafafb9

**Comment by aminya** [comment]
So this one:
```diff
❯ git diff
diff --git a/packages/optimizers/cssnano/package.json b/packages/optimizers/cssnano/package.json
index 597fe8cbf..44a02c47c 100644
--- a/packages/optimizers/cssnano/package.json
+++ b/packages/optimizers/cssnano/package.json
@@ -22,7 +22,6 @@
   "dependencies": {
     "@parcel/plugin": "2.0.0-beta.3.1",
     "@parcel/source-map": "2.0.0-rc.4",
-    "@parcel/utils": "2.0.0-beta.3.1",
     "cssnano": "^5.0.5",
     "postcss": "^8.3.0"
   }
diff --git a/packages/optimizers/cssnano/src/CSSNanoOptimizer.js b/packages/optimizers/cssnano/src/CSSNanoOptimizer.js
index bb96edebb..b2ed83fa6 100644
--- a/packages/optimizers/cssnano/src/CSSNanoOptimizer.js
+++ b/packages/optimizers/cssnano/src/CSSNanoOptimizer.js
@@ -5,15 +5,22 @@ import {Optimizer} from '@parcel/plugin';
 import postcss from 'postcss';
 import cssnano from 'cssnano';
 import type {CSSNanoOptions} from 'cssnano'; // TODO the type is based on cssnano 4
-import {loadConfig} from '@parcel/utils';
 import path from 'path';

 export default (new Optimizer({
+  async loadConfig({ config, options }) {
+    return ((await config.getConfigFrom(
+      path.join(options.entryRoot, 'index.css'),
+      ['.cssnanorc', 'cssnano.config.json', 'cssnano.config.js'],
+    )) ?? {}: CSSNanoOptions);
+  },
+
   async optimize({
     bundle,
     contents: prevContents,
     getSourceMapReference,
     map: prevMap,
+    config,
     options,
   }) {
     if (!bundle.env.shouldOptimize) {
@@ -26,14 +33,7 @@ export default (new Optimizer({
       );
     }

-    const userConfig = ((await loadConfig(
-      options.inputFS,
-      path.join(options.entryRoot, 'index.css'),
-      ['.cssnanorc ', 'cssnano.config.json', 'cssnano.config.js'],
-      options.projectRoot,
-    )) ?? {}: CSSNanoOptions);
-
+    const result = await postcss([cssnano(config)]).process(prevContents, {
       // Suppress postcss's warning about a missing `from` property. In this
       // case, the input map contains all of the sources.
       from: undefined,
```

**Comment by mischnic** [comment]
Yes. And you should also remove the space in `'.cssnanorc '`

**Comment by mischnic** [comment]
Actually, you need to do call `config.setResult` instead of returning the result in `loadConfig` like here: https://github.com/parcel-bundler/parcel/blob/1ba0bd1833d785c1acf6af36e0a1d429f1ce1fb6/packages/packagers/js/src/index.js#L21

Flow should also already be complaining about that

**Comment by aminya** [comment]
> Flow should also already be complaining about that

Well, Flow complains about other things too. `config` is not an `OptimizerOpts` property. I am just following based on your comments

**Comment by aminya** [comment]
> > Flow should also already be complaining about that
> 
> Well, Flow complains about other things too. `config` is not an `OptimizerOpts` property. I am just following based on your comments

😄 My branch was out of date.

File: flow-libs/cssnano.js.flow [hunk 1]
```diff
@@ -1,13 +1,14 @@
...
 // https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/cssnano/index.d.ts
 
 declare module 'cssnano' {
+  import type { Plugin } from "postcss"
   declare interface CSSNanoOptions {
     configFile?: string;
     preset?: [string, mixed] | string;
   }
-  declare module.exports: (?CSSNanoOptions) => void;
+  declare module.exports: (?CSSNanoOptions) => Plugin;
```
Comment (aminya) [inline flow-libs/cssnano.js.flow:13]: This was based on this, and the fact that postcss gets plugins in its input.
https://github.com/cssnano/cssnano/blob/39b4cc089e8825d308df1fa8a7313d38d8f1f2f7/packages/example-cli-usage/postcss.config.js#L2

File: packages/optimizers/cssnano/src/CSSNanoOptimizer.js [hunk 2]
```diff
@@ -4,13 +4,41 @@ import SourceMap from '@parcel/source-map';
...
+      if (isJavascript) {
+        config.shouldInvalidateOnStartup();
+      }
+
+      // Resolve relative paths from config file
+      if (configFile.contents.paths) {
+        configFile.contents.paths = configFile.contents.paths.map(p =>
+          path.resolve(path.dirname(configFile.filePath), p),
+        );
+      }
```
Comment (mischnic) [inline packages/optimizers/cssnano/src/CSSNanoOptimizer.js]: The only options in the config file are `plugins ` and `preset` https://cssnano.co/docs/config-file/

Comment (aminya) [reply packages/optimizers/cssnano/src/CSSNanoOptimizer.js]: Yes, how does it affect this?

Comment (mischnic) [reply packages/optimizers/cssnano/src/CSSNanoOptimizer.js]: That part isn't needed

Comment (aminya) [reply packages/optimizers/cssnano/src/CSSNanoOptimizer.js]: I copied this part from the Sylus package. I didn't know the reasoning behind this part of the code.

```suggestion
```

**Review by devongovett — APPROVED** [review]

**Pull request merged**